### DISCUSSION TITLE
Explore how to integrate the new View Transitions API

### DISCUSSION THREAD

**Discussion by luisherranz** [WordPress/gutenberg#53063] [Interactivity API]
Let's explore how to integrate the new [View Transitions API](https://developer.chrome.com/docs/web-platform/view-transitions/)!

@darerodz played with it [in this branch](https://github.com/WordPress/block-interactivity-experiments/blob/403a19576309d89bb86fe776c7a0bb599cc3b7f3/src/runtime/transitions.js), but everything was hardcoded. Let's see what tools we can build to make using this API a breeze for theme creators.

---

About running `document.startViewTransition` with client-side navigation:

- Should the router support it?

  - Should it be configured using the router's state? Like:

    ```js
    store({
      state: {
        router: {
          viewTransitions: true,
        },
      },
    });
    ```

  - Should it be a `<meta>` tag like the client-side navigation?

    ```html
    <meta itemprop="wp-view-transitions" content="active">
    ```

  - Should it be the same `<meta>` tag as the client-side navigation?

    ```html
    <meta itemprop="wp-interactivity" content='{ "clientSideNavigation": true, "viewTransitions": true }'>
    ```

  - Should the client-side navigation be a state configuration instead of a `<meta>` tag?

    ```js
    store({
      state: {
        router: {
          clienSideNavigation: true,
          viewTransitions: true
        },
      },
    });
    ```

  - Can we make it so `document.startViewTransition` is not hardcoded into the router, and instead is injected as external functionality? (make the router "extensible")

About setting `view-transition-name` dynamically:

- Can we create a directive to set it dynamically right before the user navigates?

  ```html
  <a
    class="wp-block-group"
    data-wp-link
    data-wp-view-transition-name="featured-image"
  >
    <h3 class="wp-block-header">Some Post Title</h3>
    <img class="wp-block-image" src=".../featured-image.jpg" />
  </a>
  ```

- Should it be part of `wp-link'?

  ```html
  <a
    class="wp-block-group"
    data-wp-link='{ "prefetch": true, "viewTransitionName": "featured-image" }'
  >
    <h3 class="wp-block-header">Some Post Title</h3>
    <img class="wp-block-image" src=".../featured-image.jpg" />
  </a>
  ```

- If this needs to match the CSS of the theme, how can a theme "inject" this directive in the correct place of the HTML?
- Should blocks support it by default? (configurable in the Editor)

---

Other questions?

**Comment by luisherranz** [comment]
@SantosGuillamot started this exploration:
- https://github.com/WordPress/block-interactivity-experiments/pull/241

**Comment by luisherranz** [comment]
The Astro team is working on an API to do client-side navigation. It has directives for the transitions, so we should take a look and check if some of their ideas are good and could be applied here as well: https://github.com/withastro/roadmap/pull/607

**Reply by luisherranz** [reply]
This API is interesting. Not to copy, but to see what custom requirements Astro devs had which led to this API: https://docs.astro.build/en/guides/view-transitions/#building-a-custom-swap-function

**Comment by luisherranz** [comment]
SvelteKit is also working on it, so it'd be interesting to explore: https://svelte.dev/blog/view-transitions

**Comment by luisherranz** [comment]
More people join the party. Now Remix and React Router: https://twitter.com/remix_run/status/1713967836798886360

**Comment by maurodf0** [comment]
Hi, there's some news about this one, seems very interesting!

**Comment by luisherranz** [comment]
React is adding support for View transitions with a <ViewTransition /> component that integrates with all the async features like Suspense and startTransition.

https://github.com/facebook/react/pull/31975

**Reply by luisherranz** [reply]
This is really good: https://react.dev/reference/react/ViewTransition

If React releases this component and Preact adds support for it, it should be easy to add a directive that uses it.

**Reply by luisherranz** [reply]
This article is also interesting: [Bringing React's <ViewTransition> to vanilla JS](https://plainvanillaweb.com/blog/articles/2025-06-12-view-transitions/)

**Comment by silaskoehler** [comment]
Now that the View Transition Plugin from the Performance Team is available (https://felix-arntz.me/blog/introducing-view-transitions-wordpress-plugin/), I was wondering if there are any updates related to this topic. I'm just not quite seeing the connection at the moment, so I was curious to ask. Was the plugin developed separately from the interactivity API?
Thanks a lot for helping me understand it better!

**Reply by westonruter** [reply]
The plugin was developed separately from the Interactivity API, yes. It was specifically developed for cross-document view transitions, being applicable to both block themes and classic themes. It is not integrated with the experimental client-side navigation with the Interactivity API.
